src/ql/pass/map/qubits/map/detail/mapper.cc

```diff
@@ -374,740 +205,226 @@ Alter Mapper::tie_break_alter(List<Alter> &alters, Future &future) {
-        // non-nearest-neighbor, otherwise they might also be nearest-neighbor).
-        gates = av_gates;
-
-#ifdef MULTI_LINE_LOG_DEBUG
-        QL_IF_LOG_DEBUG {
-            for (const auto &gate : gates) {
-                QL_DOUT(". map_mappable gates, 2q gates returned: " << gate->qasm());
-            }
+        if (!hasMappedGate) {
+            // Remaining gates require actual routing.
+            return available_gates;
         }
-#else
-        QL_DOUT("map_mappable gates, 2q gates returned (disabled)");
-#endif
-
-        return true;
     }
+    
+    return {};
 }
 
-/**
- * Select an Alter based on the selected heuristic.
- *
- *  - If BASE[_RC], consider all Alters as equivalent, and thus apply the
- *    tie-breaking strategy to all.
- *  - If MIN_EXTEND[_RC], consider Alters with the minimal cycle extension
- *    of the given past (or some factor of that amount, ordered by
- *    increasing cycle extension) and recurse. When the recursion depth
- *    limit is reached, apply the tie-breaking strategy.
- *
- * For recursion, past is the speculative past, and base_past is the past
- * we've already committed to, and should thus measure fitness against.
- */
-void Mapper::select_alter(
+
+Alter Mapper::select_alter(
     List<Alter> &alters,
-    Alter &result,
     Future &future,
     Past &past,
     Past &base_past,
     UInt recursion_depth
 ) {
-    // alters are all alternatives we enter with. There must be at least one.
     QL_ASSERT(!alters.empty());
 
-    QL_DOUT("select_alter ENTRY level=" << recursion_depth << " from " << alters.size() << " alternatives");
-
-    // Handle the basic strategy, where we just tie-break on all alters without
-    // recusing.
     if (options->heuristic == Heuristic::BASE || options->heuristic == Heuristic::BASE_RC) {
-        Alter::debug_print(
-            "... select_alter base (equally good/best) alternatives:", alters);
-        result = tie_break_alter(alters, future);
-        result.debug_print("... the selected Alter is");
-        // QL_DOUT("SelectAlter DONE level=" << level << " from " << la.size() << " alternatives");
-        return;
+        return tie_break_alter(alters, future);
     }
 
-    QL_ASSERT(
-        options->heuristic == Heuristic::MIN_EXTEND ||
-        options->heuristic == Heuristic::MIN_EXTEND_RC ||
-        options->heuristic == Heuristic::MAX_FIDELITY
-    );
+    QL_ASSERT(options->heuristic == Heuristic::MIN_EXTEND || options->heuristic == Heuristic::MIN_EXTEND_RC);
 
-    // Compute a score for each alternative relative to base_past, and sort the
-    // alternatives based on it, minimum first.
     for (auto &a : alters) {
-        a.debug_print("Considering extension by alternative: ...");
-        a.extend(past, base_past);           // locally here, past will be cloned and kept in alter
-        // and the extension stored into the a.score
+        a.extend(past, base_past); // This fills a.score.
```

Comment:
Recently reviewing the complexity of the mapper, we found out that 'past' is about to be copied here because the 1st formal parameter of a.extend is not a reference parameter. This copying is only needed to compute the score of the alternative from. The score computation doesn't need the (extended) list of output_gates in the copy, just the FreeCycle map.
So I propose to delete this list of gates copying from this Past copying somehow.
Also as far as I can judge, is the waiting_gates list not used.
The output_gates list is only used for the base_past to store the committed gates in.
This deletion of copying the gates list, reduces the space and time complexity of minextend by one order.

Reply:
#486